synthesis/synthesis.go

```diff
@@ -0,0 +1,388 @@
+/*
+Package synthesis fixes synthetic DNA molecules in preparation for synthesis.
+
+Many synthesis companies have restrictions on the DNA they can synthesize. This
+synthesis fixer takes advantage of synonymous codons in protein coding
+sequences (CDS) to remove problematic sequences that either users don't want
+(like restriction enzymes sites) or that would cause DNA synthesis companies to
+reject a synthesis project.
+
+This synthesis fixer is meant to cover the majority of use cases for DNA
+fixing. It is not intended to cover all possible use cases, since the majority
+of DNA design does not actually have these edge cases.
+
+FixCds does not guarantee that all requested features will be removed. If you
+have use case that FixCds cannot properly fix, please put an issue in the poly
+github.
+*/
+package synthesis
+
+import (
+	"errors"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/TimothyStiles/poly/checks"
+	"github.com/TimothyStiles/poly/transform"
+	"github.com/TimothyStiles/poly/transform/codon"
+	"github.com/jmoiron/sqlx"
+	_ "modernc.org/sqlite" // imports CGO-less sqlite
+)
+
+// FixIterations is the standard default for the number of rounds FixCds will
+// iterate for while trying to find a fixed solution for a sequence.
+var FixIterations = 100
+
+// DnaSuggestion is a suggestion of a fixer, generated by a
+// problematicSequenceFunc.
+type DnaSuggestion struct {
+	Start          int    `db:"start"`
+	End            int    `db:"end"`
+	Bias           string `db:"gcbias"`
+	QuantityFixes  int    `db:"quantityfixes"`
+	SuggestionType string `db:"suggestiontype"`
+}
+
+// Change is a change to a given DNA sequence. A list of changes is given as
+// the output of FixCds.
+type Change struct {
+	Position int    `db:"position"`
+	Step     int    `db:"step"`
+	From     string `db:"codonfrom"`
+	To       string `db:"codonto"`
+	Reason   string `db:"reason"`
+}
+
+type dbDnaSuggestion struct {
+	Start          int    `db:"start"`
+	End            int    `db:"end"`
+	Bias           string `db:"gcbias"`
+	QuantityFixes  int    `db:"quantityfixes"`
+	SuggestionType string `db:"suggestiontype"`
+	Step           int    `db:"step"`
+	ID             int    `db:"id"`
+}
+
+// RemoveSequence is a generator for a problematicSequenceFuncs for specific
+// sequences.
+func RemoveSequence(sequencesToRemove []string, reason string) func(string, chan DnaSuggestion, *sync.WaitGroup) {
+	return func(sequence string, c chan DnaSuggestion, wg *sync.WaitGroup) {
+		var enzymes []string
```

Comment:
It's not so important but, sequencesToRemove instead of enzymes is a better variable name than enzymes. I know, normally people will use this for removing restriction enzymes, but I think the purpose of the function is to allow users to remove more than restriction enzyme patterns.

Reply:
Fixed.